### PR TITLE
JENKINS-48674: downloads image for newer version from download.docker.com

### PR DESCRIPTION
**Pull request by denisa**
Update the plugin to use the new host and url scheme while preserving access to older images (pre 17.05.0-ce).

https://issues.jenkins-ci.org/browse/JENKINS-48674

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/docker/commons/tools/DockerToolInstaller.java [hunk 1]
```diff
@@ -142,6 +146,27 @@ public FilePath performInstallation(ToolInstallation toolInstallation, @Nonnull
...
+        String osName="";
+        if (os.startsWith("linux")) osName = "Linux";
+        if (os.startsWith("win")) osName = "Windows";
+        if (os.startsWith("mac")) osName = "Darwin";
+        return new URL("https://get.docker.com/builds/" + osName + os.substring(os.lastIndexOf("/")) + "/docker-" + version);
+    }
+
+    private static VersionNumber parseVersion(String version) {
+        // any version that sorts before 17.05.0-ce
+        if (version.equals("latest")) return new VersionNumber("0");
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/docker/commons/tools/DockerToolInstaller.java]: I would rather prefer an API with `null` and annotations, but fine for a private method

File: src/main/java/org/jenkinsci/plugins/docker/commons/tools/DockerToolInstaller.java [hunk 2]
```diff
@@ -142,6 +146,27 @@ public FilePath performInstallation(ToolInstallation toolInstallation, @Nonnull
...
     }
 
+    static URL getDockerImageUrl(String os, String version) throws MalformedURLException {
+        if (parseVersion(version).isNewerThan(parseVersion("17.05.0-ce")))
+            return new URL("https://download.docker.com/" + os + "/docker-" + version);
+
+        String osName="";
+        if (os.startsWith("linux")) osName = "Linux";
+        if (os.startsWith("win")) osName = "Windows";
+        if (os.startsWith("mac")) osName = "Darwin";
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/docker/commons/tools/DockerToolInstaller.java]: Would it make sense to print warnings if there is no match? (e.g. new platform)

File: src/main/java/org/jenkinsci/plugins/docker/commons/tools/DockerToolInstaller.java [hunk 3]
```diff
@@ -142,6 +146,27 @@ public FilePath performInstallation(ToolInstallation toolInstallation, @Nonnull
         return install;
     }
 
+    static URL getDockerImageUrl(String os, String version) throws MalformedURLException {
+        if (parseVersion(version).isNewerThan(parseVersion("17.05.0-ce")))
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/docker/commons/tools/DockerToolInstaller.java]: `isOlderThan()`?

File: src/main/java/org/jenkinsci/plugins/docker/commons/tools/DockerToolInstaller.java (hunk 1 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/docker/commons/tools/DockerToolInstaller.java]: Would `latest` not be “newer than” `17.05.0-ce`? Your test does request this go to `get.docker.com`. Yet https://get.docker.com/builds/Linux/x86_64/docker-latest is a 403. OTOH https://download.docker.com/linux/static/stable/x86_64/docker-latest is a 404. It seems `latest` just does not work any more—the index of https://download.docker.com/linux/static/ lists `edge`, `stable`, and `test`, but the file names in each specify a version. However I will check the test behavior and see what is going on.

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/docker/commons/tools/DockerToolInstaller.java]: Ah, https://get.docker.com/builds/Linux/x86_64/docker-latest.tgz is a 200.

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/docker/commons/tools/DockerToolInstaller.java]: …and it is `17.05.0-ce`. So they are still publishing to `get.docker.com`, just not the actual “latest” (which seems to be `17.12.0-ce` currently). Weird.

File: src/test/java/org/jenkinsci/plugins/docker/commons/tools/DockerToolInstallerTest.java [hunk 4]
```diff
@@ -56,35 +74,40 @@ public void smokes() throws Exception {
...
+        try {
+            new URL("https://download.docker.com/").openStream().close();
+        } catch (IOException x) {
+            Assume.assumeNoException("Cannot contact download.docker.com, perhaps test machine is not online", x);
+        }
         r.jenkins.getDescriptorByType(DockerTool.DescriptorImpl.class).setInstallations(
             new DockerTool("latest", "", Collections.singletonList(new InstallSourceProperty(Collections.singletonList(new DockerToolInstaller("", "latest"))))),
-            new DockerTool("1.10.0", "", Collections.singletonList(new InstallSourceProperty(Collections.singletonList(new DockerToolInstaller("", "1.10.0"))))));
+            new DockerTool("1.10.0", "", Collections.singletonList(new InstallSourceProperty(Collections.singletonList(new DockerToolInstaller("", "1.10.0"))))),
+            new DockerTool("17.09.1-ce", "", Collections.singletonList(new InstallSourceProperty(Collections.singletonList(new DockerToolInstaller("", "17.09.1-ce.tgz"))))));
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/docker/commons/tools/DockerToolInstallerTest.java]: The `version` here is incorrect.

**Comment by jglick** [comment]
I am working on a fixed PR.

**Pull request merged**